sabnzbd/interface.py

```diff
@@ -1107,33 +1111,34 @@ def saveGeneral(self, **kwargs):
     @secured_expose(check_api_key=True, check_configlock=True)
     def uploadConfig(self, **kwargs):
         admin_backup = kwargs.get("admin_backup")
+        valid_files = list(ADMIN_FILES)
+        valid_files.append("sabnzbd.ini")
 
-        # Only accept the backup file if it can be opened as a zip archive and contains sabnzbd.ini
+        # Only accept the backup file if it can be opened as a zip archive and only contains allowed file names
         valid_backup = False
         try:
             # Read the zip file into RAM so that it can be closed and deleted before restart
-            backup_ref = io.BytesIO(admin_backup.file.read())
-            admin_backup.file.close()
-            admin_zip_ref = zipfile.ZipFile(backup_ref, "r")
-            for f in admin_zip_ref.infolist():
-                if "/" in f.filename or "\\" in f.filename:
-                    logging.debug("Backup archive contains invalid file %s", f.filename)
-                    valid_backup = False
-                    break
-                if f.filename == "sabnzbd.ini":
-                    valid_backup = True
+            admin_backup_data = admin_backup.file.read()
+            with io.BytesIO(admin_backup_data) as backup_ref:
+                admin_backup.file.close()
+                with zipfile.ZipFile(backup_ref, "r") as admin_zip_ref:
+                    for f in admin_zip_ref.infolist():
+                        if f.filename not in valid_files:
+                            logging.debug(
+                                "Backup archive contains invalid file %s, allowed: %s", f.filename, valid_files
+                            )
+                            valid_backup = False
+                            break
+                        if f.filename == "sabnzbd.ini":
+                            valid_backup = True
         except:
             valid_backup = False
 
         if valid_backup:
-            sabnzbd.RESTORE_DATA = admin_zip_ref
-            Thread(target=sabnzbd.trigger_restart, kwargs={"timeout": 1}).start()
-            return sabnzbd.api.report(data={"result": True, "message": "Success"})
+            sabnzbd.RESTORE_DATA = admin_backup_data
```

Comment:
I changed this to a copy of the file data because the backup_ref file was being closed, probably by the garbage collector.